### DISCUSSION TITLE
Initializing on-memory database gives " Entity is not managed."

### DISCUSSION THREAD

**Discussion by ziyuang** [mikro-orm/mikro-orm#5658] [Q&A] [answered]
My on-memory SQLite database contains some singers and songs (just names as strings). There is a many-to-many relationship between the two entities. Here are the entities and config:

```typescript
// entities.ts

import { Collection, Entity, ManyToMany, PrimaryKey, Property } from "@mikro-orm/core";

@Entity()
export class Song {
  @PrimaryKey()
  _id: number;

  @Property({ index: true })
  name: string;

  @ManyToMany({ entity: () => Singer, mappedBy: (singer) => singer.songs })
  singers = new Collection<Singer>(this);

  constructor(name: string) {
    this.name = name;
  }
}

@Entity()
export class Singer {
  @PrimaryKey()
  _id: number;

  @Property({ index: true })
  name: string;

  @ManyToMany({ entity: () => Song, mappedBy: (song) => song.singers, owner: true })
  songs = new Collection<Song>(this);

  constructor(name: string) {
    this.name = name;
  }
}
```

```typescript
// init-db.ts

import { EntityManager } from "@mikro-orm/sqlite";

import { Singer, Song } from "./entities";

export function seedMedia(em: EntityManager) {
  const mediaRepo = {
    singer1: ["song1", "song2"],
    singer2: ["song1", "song3"],
  };

  // scan mediaRepo and create the entities and build the many-to-many relationship
  const singerEntities: Singer[] = [];
  const songEntities: { [name: string]: Song } = {};
  for (const [singerName, songs] of Object.entries(mediaRepo)) {
    const singer = em.create(Singer, { name: singerName });
    singerEntities.push(singer);
    singer.songs.init();
    for (const songName of songs) {
      let song: Song;
      if (songName in songEntities) {
        song = songEntities[songName];
      } else {
        song = em.create(Song, { name: songName });
        song.singers.init();
        songEntities[songName] = song;
      }
      singer.songs.add(song);
      song.singers.add(singer);
    }
  }
  for (const singer of singerEntities) {
    em.persist(singer);
  }
  for (const song of Object.values(singerEntities)) {
    em.persist(song);
  }
}
```

```typescript
// mikro-orm.config.ts

import { TsMorphMetadataProvider } from "@mikro-orm/reflection";
import { defineConfig } from "@mikro-orm/sqlite";

import { Singer, Song } from "./entities";

export default defineConfig({
  entities: [Singer, Song],
  entitiesTs: [Singer, Song],
  dbName: ":memory:",
  metadataProvider: TsMorphMetadataProvider,
});
```
```typescript
// app.module.ts

import { MikroOrmModule } from "@mikro-orm/nestjs";
import { Module } from "@nestjs/common";

import mikroOrmConfig from "./mikro-orm.config";
import { Singer } from "./entities";

@Module({
  imports: [MikroOrmModule.forFeature([Singer])],
  controllers: [],
  providers: [],
})
export class SingerModule {}

@Module({
  imports: [MikroOrmModule.forRoot(mikroOrmConfig), SingerModule],
  controllers: [],
  providers: [],
})
export class AppModule {}

```

```typescript
// main.ts

import { MikroORM } from "@mikro-orm/core";
import { NestFactory } from "@nestjs/core";

import { AppModule } from "./app.module";
import { seedMedia } from "./init-db";
import mikroOrmConfig from "./mikro-orm.config";

async function bootstrap() {
  const orm = await MikroORM.init(mikroOrmConfig);
  const fork = orm.em.fork();
  seedMedia(fork);
  fork.flush();
  fork.clear();
  const app = await NestFactory.create(AppModule);
  await app.listen(process.env.SERVER_PORT || 3000);
}

bootstrap();
```

Then if I run `nest start`, I will get the following error:
```
node_modules\@mikro-orm\core\errors.js:41
        return new ValidationError(`Entity ${entity.constructor.name} is not managed. An entity is managed if its fetched from the database or registered as new through EntityManager.persist()`);
               ^
ValidationError: Entity Singer is not managed. An entity is managed if its fetched from the database or registered as new through EntityManager.persist()
    at Function.entityNotManaged (node_modules\@mikro-orm+core@6.2.8\node_modules\@mikro-orm\core\errors.js:41:16)
    at Collection.getEntityManager (node_modules\@mikro-orm+core@6.2.8\node_modules\@mikro-orm\core\entity\Collection.js:264:44)
    at Collection.init (node_modules\@mikro-orm+core@6.2.8\node_modules\@mikro-orm\core\entity\Collection.js:214:25)
    at seedMedia (init-db.ts:15:18)
    at bootstrap (main.ts:11:12)
```

But I think every created entity has been persisted in the code. So what is the problem?

**Comment by B4nan** [comment, marked as answer]
That error should be pretty clear if you pay more attention to the stack trace. You are calling `Collection.init` on a not managed entity, you cant do that. Note that you also forgot to await this call - the method is async.

```ts
// drop those lines completely, the collections are already initialized since you just created the entities,
// also you don't even need to init a collection to be able to add to it
singer.songs.init();
song.singers.init();
```

Note that most of the code you provided here is unnecessary and can be replaced with a few `em.create` calls. That method can create a complete entity graph for you, including the relations. And that method already marks entities for persistence, there is no point in calling `em.persist` on top of it explicitly.

**Reply by B4nan** [reply]
You also don't need to do this (one is enough):

```
singer.songs.add(song);
song.singers.add(singer);
```

Since the ORM automatically ensures consistency among managed entities: 

https://mikro-orm.io/docs/propagation

**Reply by B4nan** [reply]
Note that what you are trying to do seems also weird, what about existing songs in the database with the same name? Your payload should have their PKs instead of names. Then it would be also much simpler to create the entity graph out of that.

**Reply by ziyuang** [reply]
> also you don't even need to init a collection to be able to add to it

I got the idea from the [official example](https://mikro-orm.io/docs/collections#working-with-collections):
```typescript
const author = em.findOne(Author, '...', { populate: ['books'] }); // populating books collection
await author.books.init();
// later
author.books.add(book);
```
where `init` is used. So what is the difference between the `author`s returned by `em.find` and `em.create`?

> there is no point in calling em.persist on top of it explicitly.

Isn't the `persistOnCreate` option set to `false` [by default](https://mikro-orm.io/docs/guide/project-setup#seeding-the-database)?

> https://mikro-orm.io/docs/propagation

Good information, thanks.

> Your payload should have their PKs instead of names.

In reality `mediaRepo` comes from a scan to a folder so there is no primary key. But yes I need to take care of songs with the same name.

---

Thanks for the help. Setting aside the duplicate song name issue, my seeding code looks like the below and the previous error is gone:
```typescript
import { EntityManager } from "@mikro-orm/sqlite";

import { Singer, Song } from "./entities";

export function seedMedia(em: EntityManager) {
  const mediaRepo = {
    singer1: ["song1", "song2"],
    singer2: ["song1", "song3"],
  };
  const songEntities: { [name: string]: Song } = {};
  for (const [singerName, songs] of Object.entries(mediaRepo)) {
    const singer = em.create(Singer, { name: singerName });
    for (const songName of songs) {
      let song: Song;
      if (songName in songEntities) {
        song = songEntities[songName];
      } else {
        song = em.create(Song, { name: songName });
        songEntities[songName] = song;
      }
      singer.songs.add(song);
    }
  }
}
```
I also changed the database to a file-based one so that `mikro-orm schema:create --run` can create the database and the tables If I don't want to do it programmatically. The new problem is that `localhost:3000/singer` returns just an empty list. I have set up the controller and the service classes in a standard way. And some debug print inside `seedMedia` shows that `em.count(Singer) == 0` after the whole for loop. So what went wrong here?

**Reply by B4nan** [reply]
> I got the idea from the [official example](https://mikro-orm.io/docs/collections#working-with-collections):

That is example about loading from the database, not about creating new entity. I will change it to use `load()` method instead, that would be fine since it wont do anything on an initialized collection (which you already have for new entities).

> Isn't the persistOnCreate option set to false [by default](https://mikro-orm.io/docs/guide/project-setup#seeding-the-database)?

Nope, it's enabled by default for [quite some time](https://mikro-orm.io/api/core/changelog#550-2022-10-23), will fix the docs.

> What went wrong here?

You need to call `em.flush()`, that is missing from your `seedMedia` function.

**Reply by B4nan** [reply]
Improved the mentioned docs here 4f40b6dad311347b8098c9ec693de474b5180f6c

**Reply by ziyuang** [reply]
> You need to call `em.flush()`, that is missing from your `seedMedia` function.

Right. That's it. I had one in `bootstrap` under `main.ts` but that doesn't work; maybe one of the sync vs. async things. Thanks for the help along the way.

---

Generally I think the documentation is well-written within its own page/section, but I always feel gaps between the pages. For example when the idea of `EntityManager` is [introduced](https://mikro-orm.io/docs/guide/first-entity#working-with-entity-manager), the example immediately uses `em.persist(...)` etc. I could successfully guess `em` is an instance of `EntityManager`, but I had a hard time finding a way to get the `em` from the air in the first place. Overall, I feel that a Mikro ORM user is assumed to have experience with [an ORM library with similar concepts](https://typeorm.io/#using-entity-manager) on NodeJS, and the documentation focuses more on mapping the user's mindset on Mikro ORM's facilities. I have worked more on the Python side (Pydantic + SQLAlchemy) so gaps like this are more perceivable to me.

I also hope the API pages can be more "accessible". For example when the documentation mentions `em.persist`, it could have been [`em.persist`](https://mikro-orm.io/api/core/class/EntityManager#persist). You might say how about using the search functionality? Well you cannot really land on an API page there; the result set seems the same no matter whether you start the top-right search on the "Docs" or "API" page. For example when you mention `load()`, I had to use the tree view in the navigation panel on the left to find the class `Collection` and then reach the information of the function.

**Reply by B4nan** [reply]
> but I had a hard time finding a way to get the em from the air in the first place

I can't believe my own eyes, the `MikroORM.init()` call is not described there at all, how come I missed that 🤦 

Thanks for the pointers.
